### ISSUE TITLE
Table file compaction

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#1830]
When a node loses its responsibility as a secondary or primary for a shard, it deletes the documents from its local storage (triggered by reactor_be_nothing).
However the actual storage file (as in 'rethinkdb_data/00000000-0000-0000-0000-000000000000') is never deleted.

Even though https://github.com/rethinkdb/rethinkdb/issues/310 has been implemented, garbage collection is not aggressive enough to actually allow shrinking the file in many cases. As a consequence, huge files keep lying around on the nodes and never get cleaned up.

We could either explicitly delete the files if the table is empty (if we can do so safely), or look into what stops the file from being shrunk in practice, despite all documents having been deleted.

**Comment by phil-hildebrand** [comment]
Will this also cover issues with files that could be compacted after rebalancing shards, or reconfiguring with additional shards?

This is something we could use given how fast our tables grow (we're predicting some to grow by 7 TB this year, so adding shards as we add nodes would be easier to manage if we new the datafiles would be compacted after the rebalance

**Comment by danielmewes** [comment]
@phil-hildebrand Yeah, the compaction operation would cover that. It's not clear yet when we'll get around to implementing this though.

**Comment by danielmewes** [comment]
Renamed issue from "We never seem to delete table files" to "Table file compaction" because that more accurately describes the remaining issue.
